### ISSUE TITLE
ArrayIndexOutOfBoundsException when spark batch insert

### ISSUE THREAD

**Issue by pan3793** [housepower/ClickHouse-Native-JDBC#155]
### JDBC version in pom.xml
2.2-stable

### ClickHouse version
20.8-lts

### Spark version
2.4.0-cdh6.3.1 with scala 2.11 and Hadoop 3.0.0-cdh6.3.1

### Error  logs
```
java.lang.ArrayIndexOutOfBoundsException: Array index out of range: 1048576
	at net.jpountz.util.SafeUtils.checkRange(SafeUtils.java:24)
	at net.jpountz.util.SafeUtils.checkRange(SafeUtils.java:32)
	at net.jpountz.lz4.LZ4JavaSafeCompressor.compress(LZ4JavaSafeCompressor.java:141)
	at net.jpountz.lz4.LZ4Compressor.compress(LZ4Compressor.java:95)
	at com.github.housepower.jdbc.buffer.CompressedBuffedWriter.flushToTarget(CompressedBuffedWriter.java:63)
	at com.github.housepower.jdbc.serializer.BinarySerializer.flushToTarget(BinarySerializer.java:79)
	at com.github.housepower.jdbc.connect.PhysicalConnection.disPhysicalConnection(PhysicalConnection.java:108)
	at com.github.housepower.jdbc.ClickHouseConnection.close(ClickHouseConnection.java:57)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:710)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:834)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:834)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:935)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:935)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2121)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2121)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$11.apply(Executor.scala:407)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1408)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:413)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
20/10/19 18:29:45 ERROR executor.CoarseGrainedExecutorBackend: RECEIVED SIGNAL TERM
```

### Steps to reproduce
ClickHouse Table DDL
```
CREATE TABLE xm_user (
  create_time       DateTime('Asia/Shanghai'),
  update_time       DateTime('Asia/Shanghai') COMMENT 'keep the larger record when merge',
  org_id            String                    COMMENT 'pk_1',
  user_id           String                    COMMENT 'pk_2',
  cf001             Nullable(Float64),
  cf002             Nullable(Float64),
  cf003             Nullable(Float64),
  cf004             Nullable(Float64),
  cf005             Nullable(Float64),
  cs001             Nullable(String),
  cs002             Nullable(String),
  cs003             Nullable(String),
  cs004             Nullable(String),
  cs005             Nullable(String),
  cs006             Nullable(String),
  cs007             Nullable(String),
  cs008             Nullable(String),
  cs009             Nullable(String),
  cs010             Nullable(String),
  cs011             Nullable(String),
  cs012             Nullable(String),
  cs013             Nullable(String),
  cs014             Nullable(String),
  cs015             Nullable(String),
  ct001             Nullable(DateTime('Asia/Shanghai')),
  ct002             Nullable(DateTime('Asia/Shanghai')),
  ct003             Nullable(DateTime('Asia/Shanghai')),
  ct004             Nullable(DateTime('Asia/Shanghai')),
  ct005             Nullable(DateTime('Asia/Shanghai')),
  cl001             Nullable(Int64),
  cl002             Nullable(Int64),
  cl003             Nullable(Int64),
  cl004             Nullable(Int64),
  cl005             Nullable(Int64)
) ENGINE = ReplacingMergeTree("update_time")
ORDER BY (org_id, user_id);
```
// TODO

### Other descriptions
Some investigation.

I found lz4 has a version conflict with my Spark(2.4.0-cdh6.3.1), and spark prefer bundled version by default. And I found you provided the shaded version but not mentioned in doc. (Maybe I missed?)

Unfortunately, same error occurred after switching to shaded version.

BTW, I think we can improve the shaded package to avoid dependence conflicts. I will open another issue to discuss it.

```
java.lang.ArrayIndexOutOfBoundsException: Array index out of range: 1048576
	at com.github.housepower.repackaged.netjpountz.util.SafeUtils.checkRange(SafeUtils.java:24)
	at com.github.housepower.repackaged.netjpountz.util.SafeUtils.checkRange(SafeUtils.java:32)
	at com.github.housepower.repackaged.netjpountz.lz4.LZ4JavaSafeCompressor.compress(LZ4JavaSafeCompressor.java:141)
	at com.github.housepower.repackaged.netjpountz.lz4.LZ4Compressor.compress(LZ4Compressor.java:70)
	at com.github.housepower.jdbc.buffer.CompressedBuffedWriter.flushToTarget(CompressedBuffedWriter.java:63)
	at com.github.housepower.jdbc.serializer.BinarySerializer.flushToTarget(BinarySerializer.java:79)
	at com.github.housepower.jdbc.connect.PhysicalConnection.disPhysicalConnection(PhysicalConnection.java:108)
	at com.github.housepower.jdbc.ClickHouseConnection.close(ClickHouseConnection.java:57)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.savePartition(JdbcUtils.scala:710)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:834)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$saveTable$1.apply(JdbcUtils.scala:834)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:935)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:935)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2121)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2121)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$11.apply(Executor.scala:407)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1408)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:413)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by sundy-li** [comment]
great, i'll try to reproduce it today

**Comment by sundy-li** [comment]
@pan3793  
There is another bug with `DateTime('Asia/Shanghai')`, which will throw out `DB::ExceptionDB::Exception: Parameter for DateTime data type must be string literal. Stack trace:`.

But if I change the   `DateTime('Asia/Shanghai')` and `DateTime`, nothing went wrong. My code is:

```
public class SparkJdbcMain {
	public static void main(String[] args) throws SQLException, ClassNotFoundException {
		SparkSession spark = SparkSession
								 .builder()
								 .appName("spark-jdbc-test")
								 .enableHiveSupport()
								 .getOrCreate();

		// mocked 170 rows data
		Dataset<Row> rdd = spark.sql("SELECT country from test_table where day = '2020-10-10'");

		rdd.foreachPartition((ForeachPartitionFunction<Row>) iterator -> {
			int columnNum = 34;
			String params = Strings.repeat("?, ", columnNum);
   
                         // I have clickhouse-jdbc in this repo
			Enumeration<Driver> drivers = DriverManager.getDrivers();
			while (drivers.hasMoreElements()) {
				DriverManager.deregisterDriver(drivers.nextElement());
			}
			DriverManager.registerDriver(new ClickHouseDriver());

			Connection connection = DriverManager.getConnection("jdbc:clickhouse://127.0.0.1:9000");
			PreparedStatement
				pstmt = connection.prepareStatement("INSERT INTO xm_user(create_time, update_time, org_id, user_id, cf001,cf002,cf003,cf004,cf005,"
													+ "cs001,cs002,cs003,cs004,cs005,cs006,cs007,cs008,cs009,cs010,cs011,cs012,cs013,cs014,cs015,"
													+ "ct001,ct002,ct003,ct004,ct005,cl001,cl002,cl003,cl004,cl005) VALUES("
													+ params.substring(0, params.length() - 2) + ")");
			while(iterator.hasNext()) {
				iterator.next();
				pstmt.setTimestamp(1, new Timestamp(System.currentTimeMillis()));
				pstmt.setTimestamp(2, new Timestamp(System.currentTimeMillis()));
				pstmt.setString(3, "4444");
				pstmt.setString(4, "5555");
				pstmt.addBatch();
			}
			pstmt.executeBatch();

		});
	}
}
```

**Comment by pan3793** [comment]
@sundy-li 
Thanks for following and investigating this issue. 
I use this table to do a benchmark on clickhouse. Since the data generator is tightly couple with our private code base. I will send you a parquet file later, then use below code should reproduce this issus.
```
try {
      spark.read.load("/xxx/xxx").write
        .mode("append")
        .option("driver", "com.github.housepower.jdbc.ClickHouseDriver")
        .option("user", params.ckUsername)
        .option("password", params.ckPassword)
        .option("batchsize", params.batchSize)
        .option("isolationLevel", "NONE")
        .jdbc(params.ckJdbcUrl, table, new Properties)
    } catch {
      case ex: SQLFeatureNotSupportedException =>
        log.warn(s"ignore ${ex.getMessage}")
    }
```

**Comment by sundy-li** [comment]
@pan3793  could you try this pr #162 for your case? 
I just tests it with spark, and it works.

You can use local jar to test it.
```
 <dependency>
            <groupId>com.github.housepower</groupId>
            <artifactId>clickhouse-native-jdbc</artifactId>
            <version>2.2-stable</version>
            <scope>system</scope>
            <systemPath>xxxx/target/clickhouse-native-jdbc-shaded-2.3-stable-SNAPSHOT.jar</systemPath>

            <exclusions>
                <exclusion>
                    <groupId>net.jpountz.lz4</groupId>
                    <artifactId>lz4</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```

**Comment by sundy-li** [comment]
>  I will send you a parquet file later
Ok, wait for that.

**Comment by pan3793** [comment]
> > I will send you a parquet file later
> 
> Ok, wait for that.

Already sent at Oct 20, 2020 17:20. Please check your mailbox, topic: *ClickHouse-Native-JDBC #155 data file*

**Comment by sundy-li** [comment]
It works with that pr.

```
:) select count() from xm_user;

SELECT count()
FROM xm_user

┌─count()─┐
│   56721 │
└─────────┘

1 rows in set. Elapsed: 0.001 sec. 

```

**Comment by jiakai908** [comment]
> 1048576

i also meet this problem. how can i solve this.

**Comment by pan3793** [comment]
> i also meet this problem. how can i solve this.

It should fixed on master, you can download the code and build in local.
@sundy-li Do we have a release plan?

**Comment by sundy-li** [comment]
Yes, I have a plan to release a new version, since we have not integration tests with spark, so after #165  #166
